### PR TITLE
GDB-11163 refreshing the page in visual graph sometimes breaks access to autocomplete

### PR DESCRIPTION
**Pull request by boyan-tonchev**
## What
In some cases, refreshing the visual graph view breaks access to autocomplete.

## Why
The implementation for loading namespaces and checking if autocomplete is enabled was inconsistent. For instance, there was a check for autocomplete status, followed by namespace loading, and then another check for autocomplete status upon successful namespace loading. Depending on the page load order, this could lead to situations where namespaces and autocomplete status were not set correctly.

## How
Introduced a global context to manage whether autocomplete is enabled and the namespaces of the selected repository. These fields are now loaded when the page is initially loaded and when the repository changes. Autocomplete and namespace loading have been separated to prevent conflicts.

## Checklist
- [x] Branch name
- [x] Target branch
- [x] Commit messages
- [x] MR name
- [x] MR Description
- [ ] Tests

### CODE REVIEW

File: src/js/angular/core/services/rdf4j-repositories.service.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import 'angular/rest/rdf4j.repositories.rest.service';
+import {namespaceListModelMapper} from "../../rest/mappers/namespaces-mapper";
+
+angular
+    .module('graphdb.framework.rdf4j.repositories.service', ['graphdb.framework.rest.rdf4j.repositories.service'])
```
Comment (svilenvelikov) [inline src/js/angular/core/services/rdf4j-repositories.service.js:5]: why this is in the rdf4j namespace

File: src/js/angular/core/services/autocomplete.service.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+angular
+    .module('graphdb.framework.core.services.autocomplete', [])
+    .factory('AutocompleteService', AutocompleteService);
+
+AutocompleteService.$inject = ['AutocompleteRestService', 'LSKeys', 'LocalStorageAdapter'];
```
Comment (svilenvelikov) [inline src/js/angular/core/services/autocomplete.service.js:5]: where is the rest service import?

File: src/js/angular/core/services/workbench-context.service.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            callback(getSelectedRepositoryNamespaces());
+        }
+        return EventEmitterService.subscribeSync(WorkbenchEventName.SELECTED_REPOSITORY_NAMESPACES_UPDATED, (payload) => callback(payload));
+    };
+
+    const resetContext = () => {
+        init();
+    };
+
+    init();
```
Comment (svilenvelikov) [inline src/js/angular/core/services/workbench-context.service.js:148]: Remove this and the function itself. The defaults can be set directly on the fields.

File: src/js/angular/core/services/workbench-context.service.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+     */
+    const onSelectedRepositoryNamespacesUpdated = (callback) => {
+        if (angular.isFunction(callback)) {
+            callback(getSelectedRepositoryNamespaces());
+        }
+        return EventEmitterService.subscribeSync(WorkbenchEventName.SELECTED_REPOSITORY_NAMESPACES_UPDATED, (payload) => callback(payload));
+    };
+
+    const resetContext = () => {
+        init();
```
Comment (svilenvelikov) [inline src/js/angular/core/services/workbench-context.service.js:145]: Remove this and do the reset here and don't forget the _selectedRepositoryId which you omitted for some reason in the init function

File: src/js/angular/core/services/workbench-context.service.js [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+     * @return {function} unsubscribe function.
+     */
+    const onSelectedRepositoryNamespacesUpdated = (callback) => {
+        if (angular.isFunction(callback)) {
+            callback(getSelectedRepositoryNamespaces());
+        }
+        return EventEmitterService.subscribeSync(WorkbenchEventName.SELECTED_REPOSITORY_NAMESPACES_UPDATED, (payload) => callback(payload));
+    };
+
+    const resetContext = () => {
```
Comment (svilenvelikov) [inline src/js/angular/core/services/workbench-context.service.js:144]: missing jsdoc here an for some of the other methods

File: src/js/angular/core/services/workbench-context.service.js [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    };
+
+    /**
+     * Updates the selected repository namespaces and emits the 'selectedRepositoryNamespacesUpdated' event to notify listeners that the selected repository namespaces are changed.
+     *
+     * @param {NamespacesListModel} selectedRepositoryNamespaces
+     */
+    const setSelectedRepositoryNamespaces = (selectedRepositoryNamespaces) => {
+        if (!isEqual(_selectedRepositoryNamespaces, selectedRepositoryNamespaces)) {
+            _selectedRepositoryNamespaces = cloneDeep(selectedRepositoryNamespaces);
```
Comment (svilenvelikov) [inline src/js/angular/core/services/workbench-context.service.js:118]: why the clone was needed here?

File: src/js/angular/controllers.js [hunk 7]
```diff
@@ -45,26 +47,34 @@ angular
...
         'graphdb.framework.utils.localstorageadapter',
         'graphdb.framework.utils.workbenchsettingsstorageservice',
-        'graphdb.framework.core.services.autocompleteStatus',
+        'graphdb.framework.core.services.autocomplete',
         'graphdb.framework.utils.uriutils',
         'graphdb.framework.guides.directives',
         'graphdb.framework.guides.services',
         'pageslide-directive',
+        'graphdb.core.services.workbench-context',
+        'graphdb.framework.rdf4j.repositories.service',
```
Comment (svilenvelikov) [inline src/js/angular/controllers.js:58]: Beware withe these services. They need to be in the core module

File: src/js/angular/controllers.js [hunk 8]
```diff
@@ -951,6 +952,45 @@ function mainCtrl($scope, $menuItems, $jwtAuth, $http, toastr, $location, $repos
...
+        if (activeRepository) {
+            RDF4JRepositoriesService.getNamespaces(activeRepository)
+                .then((namespaces) => WorkbenchContextService.setSelectedRepositoryNamespaces(namespaces));
+        } else {
+            WorkbenchContextService.setSelectedRepositoryNamespaces(new NamespacesListModel([]));
+        }
+    };
+
+    const onRepositoriesChanged = () => {
+        const activeRepository = $repositories.getActiveRepository();
```
Comment (svilenvelikov) [inline src/js/angular/controllers.js:979]: BTW, its kind of bad what we are introducing now. Currently the repositories state is held in the the repositories service which is ugly and poorly written, yes, but now we introduce an additional place where we store the selected repository and I somehow don't like this

File: src/js/angular/controllers.js [hunk 9]
```diff
@@ -951,6 +952,45 @@ function mainCtrl($scope, $menuItems, $jwtAuth, $http, toastr, $location, $repos
...
+            .then((autocompleteEnabled) => {
+                WorkbenchContextService.setAutocompleteEnabled(autocompleteEnabled);
+                LocalStorageAdapter.set(LSKeys.AUTOCOMPLETE_ENABLED, autocompleteEnabled);
+            });
+    };
+
+    const updateRepositoryNamespace = (activeRepository) => {
+        if (activeRepository) {
+            RDF4JRepositoriesService.getNamespaces(activeRepository)
+                .then((namespaces) => WorkbenchContextService.setSelectedRepositoryNamespaces(namespaces));
```
Comment (svilenvelikov) [inline src/js/angular/controllers.js:972]: Where is the catch?

**Review by svilenvelikov — CHANGES_REQUESTED** [review]

File: src/js/angular/models/namespaces/namespace.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
+export class NameSpaceModel {
```
[the remark this answers is not in the document]
Comment (boyan-tonchev) [reply src/js/angular/models/namespaces/namespace.js]: done

File: test-cypress/integration/setup/jdbc/jdbc.spec.js [hunk 11]
```diff
@@ -55,7 +55,7 @@ describe('JDBC configuration', () => {
         JdbcSteps.getJDBCConfigurations().should('contain', 'No tables are defined');
     });
 
-    it('Should create a new JDBC configuration, edit, preview, then delete', () => {
+    it.only('Should create a new JDBC configuration, edit, preview, then delete', () => {
```
Comment (svilenvelikov) [inline test-cypress/integration/setup/jdbc/jdbc.spec.js:58]: `only`

File: test-cypress/integration/setup/jdbc/jdbc.spec.js [hunk 12]
```diff
@@ -80,7 +80,7 @@ describe('JDBC configuration', () => {
         JdbcSteps.clickOnEditButton();
         // change the query,
         YasqeSteps.clearEditor();
-        YasqeSteps.writeInEditor(EDIT_QUERY);
+        YasqeSteps.pasteQuery(EDIT_QUERY);
```
Comment (svilenvelikov) [inline test-cypress/integration/setup/jdbc/jdbc.spec.js:83]: Revert these. Pasting the query doesn't trigger change and editor remains pristine which makes the tests to report false positives

**Comment by boyan-tonchev** [comment]
Jenkins retry build

File: test-cypress/integration/setup/jdbc/jdbc.spec.js [hunk 13]
```diff
@@ -80,7 +81,11 @@ describe('JDBC configuration', () => {
         JdbcSteps.clickOnEditButton();
         // change the query,
         YasqeSteps.clearEditor();
+
+        // The test fails intermittently, and the exact cause is unclear. One possible reason is that a popup appears over the cursor while typing, which may obstruct the input of the query.
+        // As a workaround, we will disable this feature by pasting the query directly and then typing a few spaces to mark the editor as modified.
         YasqeSteps.pasteQuery(EDIT_QUERY);
+        YasqeSteps.writeInEditor('  ');
```
Comment (svilenvelikov) [inline test-cypress/integration/setup/jdbc/jdbc.spec.js:88]: You could put this in the step. That's what they are used to

Comment (boyan-tonchev) [reply test-cypress/integration/setup/jdbc/jdbc.spec.js:88]: Okay, I just wanted to run the test without affecting others.

**Review by svilenvelikov — APPROVED** [review]

**Pull request merged**